src/reflect/scala/reflect/internal/util/AlmostFinalValue.java

```diff
@@ -14,93 +14,36 @@
- * parts of this implementation and special case it for `Statistics`, but that
- * caused an important performance penalty (~10%). This performance penalty is
- * due to the fact that using `static`s for the method handles and all the other
+ * to speculate on its content and effectively optimize it as if it was a constant.
+ *
+ * Originally from the JSR-292 cookbook created by Rémi Forax:
+ * https://code.google.com/archive/p/jsr292-cookbook/.
+ *
+ * Implemented in Java because using `static`s for the method handles and all the other
  * fields is extremely important for the JVM to correctly optimize the code, and
  * we cannot do that if we make `Statistics` an object extending `MutableCallSite`
- * in Scala. We instead rely on the Java implementation that uses a boxed representation.
+ * in Scala.
+ *
+ * Subsequently specialised for booleans, to avoid needless Boolean boxing.
+ *
+ * Finally reworked to default to false and only allow for the value to be toggled on,
+ * using Rémi Forax's newer "MostlyConstant" as inspiration, in https://github.com/forax/exotic.
  */
-public class AlmostFinalValue {
-  private final AlmostFinalCallSite callsite =
-      new AlmostFinalCallSite(this);
-  
-  protected boolean initialValue() {
-    return false;
-  }
-  
-  public MethodHandle createGetter() {
-    return callsite.dynamicInvoker();
-  }
-  
-  public void setValue(boolean value) {
-    callsite.setValue(value);
-  }
-  
-  private static class AlmostFinalCallSite extends MutableCallSite {
-    private Boolean value;
-    private SwitchPoint switchPoint;
-    private final AlmostFinalValue volatileFinalValue;
-    private final MethodHandle fallback;
-    private final Object lock;
-    
-    private static final Boolean NONE = null;
-    private static final MethodHandle FALLBACK;
-    static {
-      try {
-        FALLBACK = MethodHandles.lookup().findVirtual(AlmostFinalCallSite.class, "fallback",
-            MethodType.methodType(Boolean.TYPE));
-      } catch (NoSuchMethodException|IllegalAccessException e) {
-        throw new AssertionError(e.getMessage(), e);
-      }
-    }
-    
-    AlmostFinalCallSite(AlmostFinalValue volatileFinalValue) {
-      super(MethodType.methodType(Boolean.TYPE));
-      Object lock = new Object();
-      MethodHandle fallback = FALLBACK.bindTo(this);
-      synchronized(lock) {
-        value = null;
-        switchPoint = new SwitchPoint();
-        setTarget(fallback);
-      }
-      this.volatileFinalValue = volatileFinalValue;
-      this.lock = lock;
-      this.fallback = fallback;
-    }
+final class AlmostFinalValue {
+  private static final MethodHandle K_FALSE = MethodHandles.constant(boolean.class, false);
+  private static final MethodHandle K_TRUE  = MethodHandles.constant(boolean.class, true);
+
+  private final MutableCallSite callsite = new MutableCallSite(K_FALSE);
+
+  public MethodHandle createGetter() { return callsite.dynamicInvoker(); }
```

Comment:
@dwijnand Does it make a difference if you have a final `getter`? i.e.

`public final MethodHandle getter = callsite.dynamicInvoker();`

Reply:
Yeah, but does it make sense to create a new method handle manually?

You could use `COLD_STATS.getter` directly instead of `COLD_STATS_GETTER`

Reply:
In both cases it's creating a new method handle, it's just a question of where it's being stored.  Is there a reason `COLD_STATS.getter` is better than `COLD_STATS_GETTER`?

Reply:
@dwijnand Sure, but that's assuming that it is only ever called once. And if it is only expected to ever be called once, why not preallocate that instance (if the COLD_STATS_GETTER approach turns out to be faster, then using that approach still makes sense of course)?

Reply:
I'm happy to keep it as is, but let me know if you find out if preallocating it as an instance field is faster, then I'll change it.